### PR TITLE
Fs 452 merge lists

### PR DESCRIPTION
**Pull request by arkajitb**
Custom tag (!ordered_dict) not used. I am going through the contents to determine whether it is an ordered dict or not(Sequence of dictionaries).

### CODE REVIEW

**Comment by de-vri-es** [comment]
> Custom tag (!ordered_dict) not used. I am going through the contents to determine whether it is an ordered dict or not(Sequence of dictionaries).

We really can't rely on the list content for this. Doing that can lead to false positives for some data with very strange results in config loading, so we really must use a tag.

File: src/yaml.cpp [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 #include <algorithm>
 #include <cerrno>
+#include <iostream>
```
Comment (hsahin) [inline src/yaml.cpp:8]: This is redundant.

File: include/dr_param/yaml.hpp [hunk 2]
```diff
@@ -114,6 +114,7 @@ inline YamlResult<void> mergeYamlNodes(YAML::Node && map_a, YAML::Node map_b) {
 	return mergeYamlNodes(map_a, map_b);
 }
 
+}
```
Comment (hsahin) [inline include/dr_param/yaml.hpp:117]: Is this breaking anything?

File: test/yaml.cpp [hunk 3]
```diff
@@ -39,28 +39,20 @@ TEST_CASE("yaml node conversions", "[yaml_node]") {
...
-	REQUIRE(map_a["list"][0].as<int>() == 5);
-	REQUIRE(map_a["movie"].as<std::string>() == "book");
-}
-
-TEST_CASE("merge yaml nodes recursive", "[yaml_node]") {
-	YAML::Node map_a = YAML::Load("{name: aap, sub: {list: [1 , 2, 3], year: 2020}}");
-	YAML::Node map_b = YAML::Load("{sub: {list: [5], year: 2019}}");
-
+TEST_CASE("merge nodes recursive", "[yaml_node]") {
+	YAML::Node map_a = YAML::Load("{name: alpro, sub: {list: [1 , 2, 3], year: 2020}, l2: !ordered_dict[{q: 1},{w: 4}], l3: [1,2,3]}");
```
Comment (hsahin) [inline test/yaml.cpp:43]:
```suggestion
	YAML::Node map_a = YAML::Load("{name: alpro, sub: {list: [1 , 2, 3], year: 2020}, l2: !ordered_dict [{q: 1}, {w: 4}], l3: [1, 2, 3]}");
```

File: test/yaml.cpp [hunk 4]
```diff
@@ -39,28 +39,20 @@ TEST_CASE("yaml node conversions", "[yaml_node]") {
...
+	YAML::Node map_b = YAML::Load("{name: oatly, sub: {list: [5], year: 2019}, l2: !ordered_dict [{q: 2}, {w: 5}, {a: 8}], l3: [4]}");
 	auto merged = mergeYamlNodes(map_a, map_b);
 	REQUIRE(merged);
-	REQUIRE(map_a["name"].as<std::string>() == "aap");
+	REQUIRE(map_a["name"].as<std::string>() == "oatly");
 	REQUIRE(map_a["sub"]["list"].size() == 1);
 	REQUIRE(map_a["sub"]["list"][0].as<int>() == 5);
 	REQUIRE(map_a["sub"]["year"].as<int>() == 2019);
+	REQUIRE(map_a["l2"][0]["q"].as<int>()  == 2);
```
Comment (hsahin) [inline test/yaml.cpp:51]: Can you make a test case where the dict is present in node a, but not overwritten?

**Review by hsahin — CHANGES_REQUESTED** [review]
LGTM.
It would be nice to have more testcases. Like a dict value inside the ordered dict ,with multiple fields.

**Comment by arkajitb** [comment]
Is it allowed to have a dict value with multiple fields inside the ordered dict?
As per my understanding the ordered dict should only have a collection of one key value pairs.

**Comment by hsahin** [comment]
> Is it allowed to have a dict value with multiple fields inside the ordered dict?
> As per my understanding the ordered dict should only have a collection of one key value pairs.

Yes that is correct, what I meant was that the value could be a dict with multiple key value pairs.

File: src/yaml.cpp [hunk 5]
```diff
@@ -106,31 +107,49 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path) {
...
-	if (!map_b.IsMap() && !map_b.IsNull()) {
-		return YamlError{"tried to merge from a YAML node that is not a map"};
-	}
-
 	for (YAML::const_iterator iterator = map_b.begin(); iterator != map_b.end(); iterator++) {
+		
 		std::string key = iterator->first.as<std::string>();
 		YAML::Node value = iterator->second;
-		if (value.IsMap() && map_a[key]) {
+	    if (value.IsMap() && map_a[key]) {
```
Comment (de-vri-es) [inline src/yaml.cpp:114]: This is indented with spaces :scream:

I noticed a few more whitespace errors, so could you configure your editor to display whitespace? And be sure to listen to git if it tells you your commit has whitespace errors (and fix them :p ).

File: src/yaml.cpp [hunk 6]
```diff
@@ -106,31 +107,49 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path) {
...
 			auto merged = mergeYamlNodes(map_a[key], value);
 			if (!merged) {
 				merged.error().appendTrace({key, "", YAML::NodeType::Map});
 				return merged;
 			}
-			continue;
+			
+		}
+		// Check the tag of the Node and go through the contents if it is an ordered dictionary.
+		if (value.Tag() == "!ordered_dict") {
```
Comment (de-vri-es) [inline src/yaml.cpp:124]: Why check the tag on the value? I would expect you'd check the tag on `map_a` instead of each child while looping. That way, it also works when the top-level thing is an ordered dict, and you have less special cases.

You could even have `mergeYamlNodes` check it, and then delegate to `mergeYamlMaps` or `mergeYamlOrderedDicts` (or give an error). That could help keep the code easy to follow.

**Review by de-vri-es — COMMENTED** [review]
Looks good in general, some inline comments.

And one more high level comment: I think we really need to verify the YAML structure when merging. So we should check that either both nodes are maps, or if the tag is used that both are lists of single-item dicts. I don't want config errors to be silently ignored :)

File: src/yaml.cpp (hunk 6 above)
Comment (de-vri-es) [reply src/yaml.cpp:124]: As discussed on slack: it is necessary to check if the value is a map or ordered dict, because if it is not it has to be replaced rather than merged. But both maps and ordered dicts need to be merged when they're a child.

File: include/dr_param/yaml.hpp (hunk 2 above)
Comment (arkajitb) [reply include/dr_param/yaml.hpp:126]: This looks fine since the other test cases are not effected.

File: src/yaml.cpp (hunk 1 above)
Comment (arkajitb) [reply src/yaml.cpp]: removed.

File: test/yaml.cpp (hunk 3 above)
Comment (arkajitb) [reply test/yaml.cpp]: Updated.

File: test/yaml.cpp (hunk 4 above)
Comment (arkajitb) [reply test/yaml.cpp]: Updated.

File: src/yaml.cpp (hunk 5 above)
Comment (arkajitb) [reply src/yaml.cpp]: Updated.

File: src/yaml.cpp (hunk 6 above)
Comment (arkajitb) [reply src/yaml.cpp]: I have made the changes as per our discussion:)

File: include/dr_param/yaml.hpp [hunk 7]
```diff
@@ -109,11 +109,21 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path);
 /// Merge the nodes of map_b into map_a and overwrite if it already exists in map_a.
 YamlResult<void> mergeYamlNodes(YAML::Node & map_a, YAML::Node map_b);
 
-/// Merge the nodes of map_b into map_a and overwrite if it already exists in map_a.
-inline YamlResult<void> mergeYamlNodes(YAML::Node && map_a, YAML::Node map_b) {
-	return mergeYamlNodes(map_a, map_b);
+/// Merge Yaml Nodes of type Ordered Dictionary, i.e., sequence on Maps.
+YamlResult<void> mergeYamlOrderedDict(YAML::Node & map_a, YAML::Node map_b);
```
Comment (de-vri-es) [inline include/dr_param/yaml.hpp:113]: I don't think these new functions should be in the public header. We only need them to implement `mergeYamlNodes`, so the public API should be unaffected. Could you remove the new declarations and restore the old one that was removed?

File: src/yaml.cpp [hunk 8]
```diff
@@ -105,31 +103,66 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path) {
...
+
+		if (value.IsMap() && map_a[key] && map_a[key].IsMap()) {
+			auto merged = mergeYamlMaps(map_a[key], value);
 			if (!merged) {
 				merged.error().appendTrace({key, "", YAML::NodeType::Map});
 				return merged;
 			}
 			continue;
 		}
+		else if (value.Tag() == "!ordered_dict" && map_a[key]) {
```
Comment (de-vri-es) [inline src/yaml.cpp:120]: This tag should not be required on the items in `map_b`, but on the items in `map_a`. That way, the overwriting config file is cleaner, and we can hide possibly confusing tags from public config files.

File: src/yaml.cpp [hunk 9]
```diff
@@ -105,31 +103,66 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path) {
...
+			auto merged = mergeYamlOrderedDict(map_a[key], value);
+			if(!merged) {
```
Comment (de-vri-es) [inline src/yaml.cpp:122]:
```suggestion
			if (!merged) {
```

File: src/yaml.cpp [hunk 10]
```diff
@@ -105,31 +103,66 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path) {
...
+				merged.error().appendTrace({key, "", YAML::NodeType::Map});
+				return merged;
+			}
+			continue;
+		}
 		map_a[key] = value;
```
Comment (de-vri-es) [inline src/yaml.cpp:128]: Not new, but if we use an `else` here we can remove all the `continue` statements before. Not super important though, either way is fine.

```suggestion
		} else {
			map_a[key] = value;
		}
```

File: src/yaml.cpp [hunk 11]
```diff
@@ -105,31 +103,66 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path) {
...
-	// Check if the arguments are maps.
-	if (!map_a.IsMap() && !map_a.IsNull()) {
-		return YamlError{"tried to merge into a YAML node that is not a map"};
-	}
-	if (!map_b.IsMap() && !map_b.IsNull()) {
-		return YamlError{"tried to merge from a YAML node that is not a map"};
-	}
-
+// Merge Yaml Nodes (Type: Map).
+YamlResult<void> mergeYamlMaps(YAML::Node & map_a, YAML::Node map_b) {
```
Comment (de-vri-es) [inline src/yaml.cpp:107]: We should still validate the type of both nodes. I don't think that is done anymore now.

File: src/yaml.cpp [hunk 12]
```diff
@@ -105,31 +103,66 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path) {
...
+	else if (map_b.Tag() == "!ordered_dict") {
+		mergeYamlOrderedDict(map_a, map_b);
+	}
+	else if (map_b.IsNull()){
+		return estd::in_place_valid;
+	}
+	else {
+		return YamlError("map_b is neither a Map nor an Ordered dictionary");
+	}
+	return estd::in_place_valid;
```
Comment (de-vri-es) [inline src/yaml.cpp:162]: Same as before: we should check the type/tag on `node_a`, not on `node_b` to determine what we want to do. And then the type/structure of `node_b` still needs to be validated, but that makes more sense to do in the specialized functions.

Comment (de-vri-es) [reply src/yaml.cpp:162]: Please also make sure to follow the Fizyr style guide, so no line breaks between `}` and `else`.

File: src/yaml.cpp [hunk 13]
```diff
@@ -105,31 +103,66 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path) {
...
+			}
+			continue;
+		}
 		map_a[key] = value;
 	}
 	return estd::in_place_valid;
 }
 
+// Merge Yaml Nodes (Type: Ordered Dictionary).
+YamlResult<void> mergeYamlOrderedDict(YAML::Node & map_a, YAML::Node map_b) {
```
Comment (de-vri-es) [inline src/yaml.cpp:134]: This function should verify the structure of both `map_a` and `map_b`: they need to be lists of single-item maps.

File: src/yaml.cpp [hunk 14]
```diff
@@ -105,31 +103,66 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path) {
...
+	for (std::size_t i = 0; i < map_b.size(); i++) {
+		for (std::size_t j = 0; j < map_a.size(); j++){
```
Comment (de-vri-es) [inline src/yaml.cpp:136]: This now merges every item of `map_b` into every item of `map_b`. That's not exactly what it is supposed to do :p

File: include/dr_param/yaml.hpp (hunk 2 above)
Comment (de-vri-es) [reply include/dr_param/yaml.hpp:126]: Hmm, I'm confused. How can adding an unmatched closing brace not break the header?

Comment (arkajitb) [reply include/dr_param/yaml.hpp]: sorry. I was checking the wrong file. I have updated it now.

File: src/yaml.cpp (hunk 14 above)
Comment (arkajitb) [reply src/yaml.cpp]: I have changed the logic.

File: src/yaml.cpp [hunk 15]
```diff
@@ -104,30 +103,107 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path) {
...
-			auto merged = mergeYamlNodes(map_a[key], value);
-			if (!merged) {
-				merged.error().appendTrace({key, "", YAML::NodeType::Map});
-				return merged;
+// Merge Yaml Nodes (Type: Ordered Dictionary).
+YamlResult<void> mergeYamlOrderedDict(YAML::Node & map_a, YAML::Node map_b) {
+	std::size_t i = 0;
+	std::size_t j = 0;
+	std::size_t a_pointer = 0;
+	std::size_t b_pointer = 0;
```
Comment (de-vri-es) [inline src/yaml.cpp:154]: It is very hard to follow what is happening with the indices now.

I would propose to add some utility functions to make the code easier to reason about. In particular, I think it would help to have at least these helper functions:
* `std::optional<std::size_t> orderedDictFind(YAML::Node parent, std::string const & name)` 
* `void orderedDictUpdateOrInsert(YAML::Node parent, std::string const & name, YAML::Node contents)`

I think those functions will be much easier to verify on their own, and then the merge algorithm should become a lot easier to read too.

**Comment by de-vri-es** [comment]
I would still prefer if you could implement the feedback from https://github.com/fizyr/dr_param/pull/9#discussion_r624960727. That would help a lot to keep the code more readable.

**Comment by arkajitb** [comment]
I have added one helper function to keep a track of the nodes which are already found. I am not able to find a way to add the mentioned utility functions. I need some help:P

File: src/yaml.cpp [hunk 16]
```diff
@@ -147,21 +147,31 @@ YamlResult<void> mergeYamlMaps(YAML::Node & map_a, YAML::Node map_b) {
...
 
 // Uitility function to check whether a value is present in a vector.
-bool Contains(const std::vector<int> &list, int x)
-{
+bool Contains(const std::vector<int> &list, int x){
 	return std::find(list.begin(), list.end(), x) != list.end();
 }
 
+// Utility function to check whether key in map_b is present in map_a.
+bool orderedDictFind(YAML::Node map_a, YAML::Node map_b, std::vector<int> & checked, int index){
```
Comment (de-vri-es) [inline src/yaml.cpp:155]: This function should only take a single ordered dict and search for a string key. It much to specialized now.

Comment (de-vri-es) [reply src/yaml.cpp]: It's not searching for a string key now. It just checks if a certain index has the string key.

I would really like you to follow the function signatures that I proposed earlier:
```c++
    std::optional<std::size_t> orderedDictFind(YAML::Node parent, std::string const & name);

    void orderedDictUpdateOrInsert(YAML::Node parent, std::string const & name, YAML::Node contents);
```

File: src/yaml.cpp [hunk 17]
```diff
@@ -104,30 +103,132 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path) {
...
+	bool single_map = true;
+	// Boolean to determine whether the element in b is present in a or not.
+	bool found = false;
+	for (std::size_t i = 0; i < map_b.size(); i++) {
+		found = false;
+		// Check whether the element in b is a single key-value pair.
+		if (!singleMap(map_b[i])){
+			return YamlError("Ordered dictionary should only contain single item map");
+		}
+		for (YAML::const_iterator iterator = map_b[i].begin(); iterator != map_b[i].end(); iterator++){
```
Comment (de-vri-es) [inline src/yaml.cpp:187]: This shouldn't be a loop now, right? You've already checked that `map_b[i]` is a single-item map, so there is only one child.

File: src/yaml.cpp [hunk 18]
```diff
@@ -104,30 +103,132 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path) {
...
+	
+	return estd::in_place_valid;
+}
+
+// Merge Yaml Nodes.
+YamlResult<void> mergeYamlNodes(YAML::Node & map_a, YAML::Node map_b) {
+	if (!map_a.IsMap() && map_a.Tag() != "!ordered_dict" && !map_a.IsNull()) {
+		return YamlError{"tried to merge into a YAML node that is neither a map nor an ordered dictionary"};
+	}
+	else if (!map_b.IsMap() && map_b.Tag() != "!ordered_dict" && !map_b.IsNull()) {
```
Comment (de-vri-es) [inline src/yaml.cpp:214]: `map_b` shouldn't be required to have the `ordered_dict` tag, but if map_a is an ordered dict, `map_b` should be verified to be a list of single-item maps.

It probably makes more sense to first check what `map_a` is, and then check the type/structure of `map_b`, while already knowing what `map_a` is.

**Comment by ekrikelis** [comment]
@de-vri-es Could you write a list of what is missing of the implentation please?

**Comment by de-vri-es** [comment]
Closing this. If we need it in the future we can start with a new implementation.
